src/main/java/de/taimos/pipeline/aws/InvokeLambdaStep.java

```diff
@@ -33,109 +33,137 @@
 		this.functionName = functionName;
-		this.payload = payload;
 	}
-
+	
 	public String getFunctionName() {
 		return this.functionName;
 	}
-
+	
 	public Object getPayload() {
 		return this.payload;
 	}
-
+	
+	@DataBoundSetter
+	public void setPayload(Object payload) {
+		this.payload = payload;
+	}
+	
 	public String getPayloadAsString() {
-		return this.toJsonString(this.payload);
+		if (this.payload != null) {
+			return this.toJsonString(this.payload);
+		}
+		return this.payloadAsString;
 	}
-
+	
+	@DataBoundSetter
+	public void setPayloadAsString(String payloadAsString) {
+		this.payloadAsString = payloadAsString;
+	}
+	
 	private String toJsonString(Object payloadObject) {
 		return (new JsonBuilder(payloadObject)).toString();
 	}
-
+	
+	public boolean isReturnValueAsString() {
+		return this.returnValueAsString;
+	}
+	
+	@DataBoundSetter
+	public void setReturnValueAsString(boolean returnValueAsString) {
+		this.returnValueAsString = returnValueAsString;
+	}
+	
 	@Extension
 	public static class DescriptorImpl extends AbstractStepDescriptorImpl {
-
+		
 		public DescriptorImpl() {
 			super(Execution.class);
 		}
-
+		
 		@Override
 		public String getFunctionName() {
 			return "invokeLambda";
 		}
-
+		
 		@Override
 		public String getDisplayName() {
 			return "Invoke a given Lambda function";
 		}
 	}
-
+	
 	public static class Execution extends AbstractSynchronousStepExecution<Object> {
-
+		
 		private static final long serialVersionUID = 1L;
-
+		
 		@Inject
 		private transient InvokeLambdaStep step;
 		@StepContextParameter
 		private transient EnvVars envVars;
 		@StepContextParameter
 		private transient TaskListener listener;
-
+		
 		@Override
 		protected Object run() throws Exception {
 			AWSLambda client = AWSClientFactory.create(AWSLambdaClientBuilder.standard(), this.envVars);
-
+			
 			String functionName = this.step.getFunctionName();
-
+			
 			this.listener.getLogger().format("Invoke Lambda function %s%n", functionName);
-
+			
 			InvokeRequest request = new InvokeRequest();
 			request.withFunctionName(functionName);
 			request.withPayload(this.step.getPayloadAsString());
 			request.withLogType(LogType.Tail);
-
+			
 			InvokeResult result = client.invoke(request);
-
+			
 			this.listener.getLogger().append(this.getLogResult(result));
 			String functionError = result.getFunctionError();
 			if (functionError != null) {
 				throw new RuntimeException("Invoke lambda failed! " + this.getPayloadAsString(result));
 			}
-			return this.getPayloadAsObject(result);
+			if (this.step.isReturnValueAsString()) {
+				return this.getPayloadAsString(result);
+			} else {
+				return this.getPayloadAsObject(result);
+			}
 		}
-
-		private Object getPayloadAsObject(InvokeResult result) {
-			return new JsonSlurper().parseText(this.getPayloadAsString(result));
+		
+		public Object getPayloadAsObject(InvokeResult result) {
+			return JSONSerializer.toJSON(this.getPayloadAsString(result));
```

Comment:
shouldn't this be `fromJSON`?

Reply:
From which class?
[toJSON()](http://json-lib.sourceforge.net/apidocs/net/sf/json/JSONSerializer.html#toJSON(java.lang.Object)) crates a `net.sf.json.JSON` object from among other things a String.
The [readJson step](https://github.com/jenkinsci/pipeline-utility-steps-plugin/blob/master/src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/json/ReadJSONStepExecution.java#L64) uses that method, too.
Subclasses of JSON implement List or Map so they behave the same as the old implementation.
I tested the new implementation with [this Jenkinsfile](https://github.com/thilomack/pipeline-aws-plugin-invoke-lambda-test/blob/master/Jenkinsfile)

Reply:
Ok, so just the naming is weird.

Reply:
Why are the methods public now?

Reply:
That would fix the architectural weirdness of the public methods. Go ahead and I will merge it.

Reply:
Great!